producer-services/src/main/scala/za/co/absa/spline/producer/service/repo/ExecutionProducerRepositoryImpl.scala

```diff
@@ -0,0 +1,98 @@
+
+@Repository
+class ExecutionProducerRepositoryImpl @Autowired()(db: ArangoDatabaseAsync) extends ExecutionProducerRepository {
+
+
+  override def insertExecutionPlan(executionPlan: ExecutionPlan)(implicit ec: ExecutionContext): Future[UUID] = {
+    val uris = referencedUris(executionPlan)
+    for {
+      existingUriKey <- queryExistingToKey(uris)
+      newUriKey <- generateNewKeys(uris, existingUriKey)
+      res <- {
+        val params = ExecutionPlanTransactionParams(executionPlan, newUriKey, newUriKey ++ existingUriKey)
+        val transaction = db.transaction(params.getInsertScript, classOf[util.HashMap[String, String]], params.getTransactionOptions).toScala
+        transaction.map(t => UUID.fromString(t.get("_key")))
+      }
+    } yield res
+  }
+
+  override def insertExecutionEvents(executionEvents: Array[ExecutionEvent])(implicit ec: ExecutionContext): Future[Array[String]] = {
+    val executionIdList = executionEvents.map(e => "\"" + e.planId.toString + "\"").mkString(", ")
+    val query = s"for ex in execution filter ex._key in [$executionIdList] return ex._key"
+    val result = ArangoImplicits.ArangoDatabaseAsyncScalaWrapper(db).query[String](query)
+    val existingExecutionIds = result.map((c: ArangoCursorAsync[String]) => c.streamRemaining().toScala.toList)
+
+    val nonExistingExecutions = existingExecutionIds.flatMap(e => Future(executionEvents.map(e => e.planId.toString).toSet -- e toArray))
+
+    existingExecutionIds.flatMap(
+      ex => {
+        val filteredExecutionEvents = executionEvents.filter(e => ex.contains(e.planId.toString))
+        val params = ExecutionEventTransactionParams(filteredExecutionEvents)
+        db.transaction(params.getInsertScript, classOf[util.HashMap[String, String]], params.getTransactionOptions)
+        nonExistingExecutions
+      }
+    )
+  }
+
+  private def generateNewKeys(uris: Seq[String], uriToExistingKey: Map[String, String]): Future[Map[String, String]] = {
+    Future((uris.toSet -- uriToExistingKey.keys).map(_ -> randomUUID.toString).toMap)
+  }
+
+  private def referencedUris(executionPlan: ExecutionPlan) = {
+    val allOperations = executionPlan.operations.reads ++ executionPlan.operations.others ++ List(executionPlan.operations.write)
+    allOperations.flatMap {
+      case r: ReadOperation => r.inputSources
+      case w: WriteOperation => Some(w.outputSource)
+      case _ => None
+    }.distinct
+  }
+
+  private def queryExistingToKey(uris: Seq[String]): Future[Map[String, String]] = {
+    val urisList = uris
+      .map(uri => "\"" + uri + "\"")
+      .mkString(", ")
+    val query = s"for ds in dataSource filter ds.uri in [$urisList] return ds"
+    val result = ArangoImplicits.ArangoDatabaseAsyncScalaWrapper(db).query[DataSource](query)
+    result.map((c: ArangoCursorAsync[DataSource]) => {
+      val dataSourceStream = c.streamRemaining().toScala
+      dataSourceStream.map(ds => ds.uri -> ds._key.get).toMap
```

Comment:
Using `get` on `Option`.

Reply:
used orNull instead but this is link to the issue https://github.com/AbsaOSS/spline/issues/273